overlord/exportstate/exportstate.go

```diff
@@ -0,0 +1,140 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package exportstate implements the manager and state aspects responsible
+// for the exporting portions of installed snaps to the system.
+package exportstate
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+// stateMapKey returns key used for indexing the map of exported snap content.
+func stateMapKey(instanceName string, rev snap.Revision) string {
+	return instanceName + "/" + rev.String()
+}
+
+// Set remembers export manifest for a particular snap revision.
+func Set(st *state.State, instanceName string, rev snap.Revision, m *Manifest) {
+	var exports map[string]*json.RawMessage
+	if err := st.Get("exports", &exports); err != nil && err != state.ErrNoState {
+		panic("internal error: cannot unmarshal exports state: " + err.Error())
+	}
+	if exports == nil {
+		exports = make(map[string]*json.RawMessage)
+	}
+	key := stateMapKey(instanceName, rev)
+	if m == nil {
+		delete(exports, key)
+	} else {
+		data, err := json.Marshal(m)
+		if err != nil {
+			panic("internal error: cannot marshal export manifest: " + err.Error())
+		}
+		raw := json.RawMessage(data)
+		exports[key] = &raw
+	}
+	st.Set("exports", exports)
+}
+
+// Get recalls export manifest of a particular snap revision.
+func Get(st *state.State, instanceName string, rev snap.Revision, m *Manifest) error {
+	*m = Manifest{}
+
+	var exports map[string]*json.RawMessage
+	if err := st.Get("exports", &exports); err != nil {
+		// This can return state.ErrNoState
+		return err
+	}
+	key := stateMapKey(instanceName, rev)
+	raw, ok := exports[key]
+	if !ok {
+		return state.ErrNoState
+	}
+	// XXX: do we need the address?
+	if err := json.Unmarshal([]byte(*raw), &m); err != nil {
+		return fmt.Errorf("cannot unmarshal export manifest: %v", err)
+	}
+	return nil
+}
+
+// exporetedVersionSymlinkPath returns the path of the current exported version symlink
+// for given snapName.
+func exportedVersionSymlinkPath(snapName string) string {
+	return filepath.Join(dirs.ExportDir, snapName, "current")
```

Comment:
how does export-set name fit into this?

Reply:
Interesting, I also forgot about this. The answer to your question is that it does not. We are switching at a higher-level, effectively what we switch is the proposed Manifest.ExportedName.